feat(mobile): Added "jump to date" functionality to the memory view

arnolicious opened the pull request:
Added a "Jump to date" Button in the bottom Info bar of the memory view.

It was somewhat mentioned [here](https://github.com/immich-app/immich/discussions/5858), but it is mainly a feature I've found myself wanting quite often.

As to the implementation, I don't have much Flutter experience, so my approach to communicate between the two components (the button in the memory view, and the Timeline list that needs to scroll) was with events, for which I've added the [event package](https://pub.dev/packages/event)

If there's a better/more "fluttery" way to do this, let me know!

And for the UI, at first I wanted to add an IconButton, however I couldn't really find a fitting Icon for this action (the closest I could come up with, was the `open_in_new` Icon), so I settled on a TextButton, and added the text to the languages I speak.

Here's a demo of what it looks like, on a very choppy emulator :D
https://github.com/immich-app/immich/assets/46051866/289750d1-d2b5-4619-8441-2c77295dd57b

arnolicious committed:
implemented jump to date from memory

arnolicious committed:
Merge branch 'main' into feat(mobile)-jump-to-asset

shenlong-tanwen commented on mobile/assets/i18n/de-DE.json:
```diff
@@ -208,6 +208,7 @@
   "image_viewer_page_state_provider_download_error": "Fehler beim Herunterladen",
   "image_viewer_page_state_provider_download_success": "Erfolgreich heruntergeladen",
   "image_viewer_page_state_provider_share_error": "Fehler beim Teilen",
+  "jump_to_date": "Zu Datum springen",
```
We don't need the keys in language files other than en-US. The other translations are handled through Localizely.

shenlong-tanwen commented on mobile/lib/modules/home/ui/asset_grid/immich_asset_grid_view.dart:
```diff
@@ -138,6 +147,7 @@ class ImmichAssetGridViewState extends State<ImmichAssetGridView> {
     }
 
     final section = widget.renderList.elements[index];
+    // debugPrint("section: ${section.date} $index");
```
Can we removed

shenlong-tanwen commented on mobile/lib/shared/services/event.service.dart:
```diff
@@ -0,0 +1,3 @@
+import 'package:event/event.dart';
+
+final scrollToDateEvent = Event<Value<DateTime>>();
```
We do not need a separate third party package for this functionality. The same can be achieved using `ValueListenable` or even `riverpod`. We can implement a event system over this to notify widgets about updates. Can you please remove the third party library and use a different implementation instead?

shenlong-tanwen reviewed (changes requested):
Thanks a lot for working on this PR. I'm sure a lot of users would find this useful. Can you please address the review comments. Also, Can we change the button to an `IconButton` instead? Choose any icon you find appropriate for now and we can change it later. `open_in_new`,`move_down` or `timeline` seems appropriate to me

arnolicious committed:
Merge branch 'main' into feat(mobile)-jump-to-asset

arnolicious committed:
Changed implementation to a ValueNotifier & fixes

martyfuhry commented on mobile/lib/modules/home/ui/asset_grid/immich_asset_grid_view.dart:
```diff
@@ -223,6 +233,22 @@ class ImmichAssetGridViewState extends State<ImmichAssetGridView> {
+    );
+    if (index != -1) {
+      // Not sure why the index is shifted, but it works. :3
+      _scrollToIndex(index + 1);
```
This is rather odd, but if it works, it works! I wonder if we shouldn't double check the position if it's the last asset in the list to avoid an overflow?

arnolicious replied:
After some testing, it seems that scrolling to an index too high just gets silently swallowed and the list scrolls to the last item anyways, but I added a check anyhow.

martyfuhry commented on mobile/lib/modules/home/ui/asset_grid/immich_asset_grid_view.dart:
```diff
@@ -237,6 +263,9 @@ class ImmichAssetGridViewState extends State<ImmichAssetGridView> {
   void initState() {
     super.initState();
     scrollToTopNotifierProvider.addListener(_scrollToTop);
+    scrollToDateNotifierProvider
```
Do you need to remove this listener on dispose?

arnolicious replied:
Oh yeah, thanks!

martyfuhry reviewed (commented):
This looks great and is super useful!

alextran1502 committed:
Merge branch 'main' of github.com:immich-app/immich into pr/arnolicious/7323

alextran1502 committed:
remove debug code

arnolicious committed:
Merge branch 'immich-app:main' into feat(mobile)-jump-to-asset

arnolicious committed:
feat(mobile):
- Added index bound checks
- Handled edge cases when scrolling to the very bottom of the grid-view
- removing the listener on dispose

alextran1502 commented:
I tested the PR and when clicking on the button to jump to timeline, it doesn't do anything 🤔

arnolicious committed:
feat(mobile): fixed debug index offset & added debug toast for scroll errors

arnolicious committed:
Merge branch 'main' into feat(mobile)-jump-to-asset

arnolicious commented:
Well that's kind of embarassing, i had left an offset in the scroll index from when i was testing... 
I hope that was the issue, but I added an Error toast for when the scrolling fails due to the index being out of bounds (for debug purposes, I'd remove that before merging)

Did the memory view at least close when you clicked the button, @alextran1502 ?

alextran1502 commented:
> Well that's kind of embarassing, i had left an offset in the scroll index from when i was testing... I hope that was the issue, but I added an Error toast for when the scrolling fails due to the index being out of bounds (for debug purposes, I'd remove that before merging)
> 
> Did the memory view at least close when you clicked the button, @alextran1502 ?

Hello, yes, when clicking on the button, it dismisses the memory view

arnolicious committed:
feat(mobile): added more debug toasts...

arnolicious committed:
Merge branch 'main' into feat(mobile)-jump-to-asset

arnolicious commented:
I'm kinda stumped, i added another toast, in case the date that the listener receives is somehow null...

Are you testing this on the demo instance by any chance?

DawidPietrykowski commented:
I tested it on my instance and it works, hides the memory view and jumps to the correct date

arnolicious committed:
Merge branch 'main' into feat(mobile)-jump-to-asset

arnolicious committed:
Merge branch 'main' into feat(mobile)-jump-to-asset

arnolicious committed:
Merge branch 'main' into feat(mobile)-jump-to-asset

arnolicious committed:
Merge branch 'main' into feat(mobile)-jump-to-asset

arnolicious committed:
Merge branch 'main' into feat(mobile)-jump-to-asset

arnolicious committed:
Merge branch 'main' into feat(mobile)-jump-to-asset

alextran1502 committed:
Merge branch 'main' of github.com:immich-app/immich into feat(mobile)-jump-to-asset

alextran1502 commented:
Hey @arnolicious, I just tested again on my prod instance. When I clicked on the link, I got this error 

![image](https://github.com/immich-app/immich/assets/27055614/01dc891b-3fef-499c-bdac-7917def68d11)

arnolicious committed:
Merge branch 'main' into feat(mobile)-jump-to-asset

arnolicious commented:
Hm, an index of -1 would indicate no item in the list matching the date was found, which shouldn't happen, since if that memory exists, it should also exist in the timeline.

The check looks right to me, at first i suspected timezones to possibly be the cause of this, but I couldn't replicate it.
(I assume your server and your phone are in the same timezone, right?)

```dart
final index = widget.renderList.elements.indexWhere(
      (e) =>
          e.date.year == date.year &&
          e.date.month == date.month &&
          e.date.day == date.day,
);
```

Is the date in the error message correct, is that the date of the memory you clicked on?
I assume that somehow the date must be incorrect, maybe it's shifted by a day, and so I'd assume either the day before or after that memory does not have any assets... Is that the case? @alextran1502

martyfuhry commented:
I'm curious, what if you group the assets on the timeline by month instead of day? Does that change the expected index in the render list?

I'm thinking about this setting, in particular. 

![Screenshot_20240322_164141.jpg](https://github.com/immich-app/immich/assets/100457/0d4fe3b8-a104-4380-bc23-ff02c623b753)

arnolicious commented:
Oh, didn't know that was a thing lol

Gonna look into it, thanks for pointing that out!

martyfuhry commented:
You should also gracefully handle the index not found case. In the ideal case, we'd hide the button entirely. But we can also show a nicely formatted error that we can't find the day in the timeline. Maybe we can do something even better?

arnolicious commented:
I'm not sure how feasible or necessary hiding the button would be, since it really shouldn't happen, once this bug is fixed. I will make the toast error more user friendly tho.

The question now is rather, what should happen when the timeline is grouped by month?
Scrolling to the month is fine, but I'm not sure how/if it's possible to then find the specific day in the month and scroll to it... 😅

arnolicious committed:
feat(mobile): scroll to month, if timeline is not grouped by days

arnolicious committed:
Merge branch 'main' into feat(mobile)-jump-to-asset

arnolicious committed:
Merge branch 'main' into feat(mobile)-jump-to-asset

alextran1502 committed:
Merge branch 'main' into feat(mobile)-jump-to-asset

alextran1502 committed:
Merge branch 'main' of github.com:immich-app/immich into feat(mobile)-jump-to-asset
